diff --git a/Makefile b/Makefile
index df87a25..93ef496 100644
--- a/Makefile
+++ b/Makefile
@@ -157,6 +157,9 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
+	$U/_sleep\
+	$U/_pingpong\
+	$U/_find\
 
 
 ifeq ($(LAB),syscall)
diff --git a/commands.gdb b/commands.gdb
new file mode 100644
index 0000000..d3265f7
--- /dev/null
+++ b/commands.gdb
@@ -0,0 +1,37 @@
+si 4
+si
+si 7
+si
+b start.c:58
+c
+p/x main
+p/x $mepc
+si
+n 10
+n 5
+n 2
+s
+fin
+b main.c:42
+c
+s
+n
+n
+n
+n
+n
+n
+p p->name
+n
+n
+p cpus[$tp]->proc->name
+b kernel/trap.c:42
+c
+p cpus[$tp]->proc->name
+n
+n
+n
+n
+p cpus[$tp]->proc->name
+n
+q
diff --git a/user/find.c b/user/find.c
new file mode 100644
index 0000000..e06b51a
--- /dev/null
+++ b/user/find.c
@@ -0,0 +1,75 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/fs.h"
+
+/*
+文件项
+struct stat {
+  int dev;
+  uint ino;
+  short type;
+  short nlink;
+  uint64 size;
+};
+目录项
+struct dirent {
+  ushort inum;
+  char name[DIRSIZ];
+};
+*/
+char *fmtname(char *path) {
+    char *p;
+
+  // Find first character after last slash.
+    for (p = path + strlen(path); p >= path && *p != '/'; p--)
+    ;
+    p++;
+
+  // Return blank-padded name.
+    return p;
+}
+
+void find(char *path,char *name){
+    char buf[512],*p;
+    int fd;
+    struct dirent de;
+    struct stat st;
+    if((fd = open(path,0))<0){ //open(path,0) 0 表示只读
+        fprintf(2,"find: cannot open %s\n",path);
+        return;
+    }
+    if(fstat(fd,&st)<0){
+        fprintf(2,"find: cannot stat %s\n",path);
+        close(fd);
+        return;
+    }
+    if(1){
+        if(strcmp(fmtname(path),name) == 0){
+            printf("%s\n",path);
+        }
+    }
+    if(st.type == T_DIR){
+        strcpy(buf,path);
+        p = buf + strlen(buf);
+        *p++ = '/';
+        while(read(fd,&de,sizeof(de)) == sizeof(de)){
+            if(de.inum == 0) continue;
+            memmove(p, de.name, DIRSIZ);
+            p[DIRSIZ] = 0;
+            if(strcmp(p,".") == 0 || strcmp(p,"..") == 0) continue;
+            find(buf,name);
+        }
+    }
+    
+    close(fd);
+}
+int main(int argc, char *argv[])
+{
+    if(argc != 3){
+        fprintf(2,"Usage: find <directory> <filename>\n");
+        exit(-1);
+    }
+    find(argv[1],argv[2]);
+    exit(0);
+}
diff --git a/user/pingpong.c b/user/pingpong.c
new file mode 100644
index 0000000..638b8c9
--- /dev/null
+++ b/user/pingpong.c
@@ -0,0 +1,33 @@
+#include "kernel/types.h"
+#include "user/user.h"
+
+int main() {
+    int p[2];
+    int p2[2];
+    pipe(p); //父写子读
+    pipe(p2); // 父读子写
+    int father_pid = getpid();
+    int child_pid = fork();
+    if (child_pid == 0) {
+        close(p[1]); //子关闭写
+        close(p2[0]); //子关闭读
+        char buf[15];
+        int n=read(p[0],buf,4);
+        buf[n]= '\0';
+        printf("%d: received %s from pid %d\n", getpid(),buf, father_pid);
+        close(p[0]);
+        write(p2[1], "pong", 4);
+        close(p2[1]);
+    } else {
+        close(p[0]); //父关闭读
+        close(p2[1]); //父关闭写
+        write(p[1], "ping", 4);
+        close(p[1]);
+        char buf[15];
+        int n=read(p2[0],buf,4);
+        buf[n] = '\0';
+        printf("%d: received %s from pid %d\n", father_pid, buf, child_pid);
+        close(p2[0]);
+    }
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/sleep.c b/user/sleep.c
new file mode 100644
index 0000000..3401695
--- /dev/null
+++ b/user/sleep.c
@@ -0,0 +1,19 @@
+#include "kernel/types.h"
+#include "user/user.h"
+
+int main(int argc, char *argv[])
+{
+    // printf("argc=%d\n",argc);
+    // for(int i=0;i<argc;i++){
+    //     printf("argv[%d]=%s\n",i,argv[i]);
+    // }
+    if(argc != 2){
+        printf("Sleep needs one argument!\n");
+        exit(-1);
+    }
+
+    int ticks = atoi(argv[1]);
+    sleep(ticks);
+    printf("Sleep %d\n", ticks);
+    exit(0);
+}
